src/support/threads.cpp

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+
+#include <iostream>
+
+#include "threads.h"
+
+
+// debugging tools
+
+#ifdef BINARYEN_THREAD_DEBUG
+static std::mutex debug;
+#define DEBUG_PRINT(x) { std::lock_guard<std::mutex> lock(debug); x; }
+#else
+#define DEBUG_PRINT(x)
+#endif
+
+
+namespace wasm {
+
+// Global thread information
+
+std::atomic_bool setMainThreadId(false);
+std::thread::id mainThreadId;
+
+struct MainThreadNoter {
+  MainThreadNoter() {
+    // global ctors are called on main thread
+    mainThreadId = std::this_thread::get_id();
+    setMainThreadId = true;
+  }
+};
+
+static MainThreadNoter mainThreadNoter;
+
+static std::unique_ptr<ThreadPool> pool;
+
+
+// Thread
+
+Thread::Thread() {
+  // main thread object's constructor itself can
+  // happen before onMainThread is ready
+  assert(onMainThread());
+  thread = std::unique_ptr<std::thread>(new std::thread(mainLoop, this));
+}
+
+Thread::~Thread() {
+  assert(onMainThread());
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    done = true;
+    condition.notify_one();
+  }
+  thread->join();
+}
+
+void Thread::work(std::function<ThreadWorkState ()> doWork_) {
+  // TODO: fancy work stealing
+  DEBUG_PRINT(std::cerr << "[Thread] send work to thread\n");
```

Comment:
Print thread id here and elsewhere? Makes it easier to de-interleave different debug prints.